Allow iterating over compiled ruleset before scanning

jpohls1 opened the pull request:
This PR fixes #135. It includes an updated `tutorial.rs` showcasing how to iterate over the compiled ruleset and disabling certain rules. This feature is crucial for implementing custom rule filters in client-side applications.

Please let me know any feedback and thanks for creating these bindings!

### Implementation details

#### On the new get_rules.c file

The approach taken by existing iterators such as `TagIterator` is difficult to apply, because the [rules_table](https://github.com/VirusTotal/yara/blob/master/libyara/include/yara/types.h#L562) is not properly exposed via bindgen. We just have a binary blob to work with:
```rust
pub struct YR_RULES {
    pub _bindgen_opaque_blob: [u64; 10usize],
}
```

Therefore, I created a new file `get_rules.c` which uses the `yr_rules_foreach` macro to retrieve the rules from the ruleset. This is inspired by the [Go bindings](https://github.com/hillu/go-yara/blob/fdb067a8659e2dd6e08ec2d9a86a8a121017a747/rule.go#L94).

 Since we need to compile this file in any case, the `cc` build-dependency is now mandatory, hence the changes in the `Cargo.toml` and `build.rs`.

#### On the new RulesetRule struct

The existing `Rule` struct represents a rule that matched during a scan. It would have involved some more refactoring to lift this assumption. I went with the less invasive change here and created a new `RulesetRule` struct.

vthib commented:
The YR_RULES type is opaque since https://github.com/Hugal31/yara-rust/commit/43d0be1cdd8bafaffa26cfb0c247da573fe1a818

It was made opaque mostly for codegen size reasons, so we can totally decide to expand it once again. This would give access to the rules_table field, which would be imho preferred, as it avoids adding C code for the bindings.

Hugal31 commented:
I am not a big fan of adding a mandatory dependence to cc. Is `yr_rules_foreach` the only public interface we got or can we safely re-implement it in Rust?

jpohls1 commented:
Cool, didn't know about 43d0be1cdd8bafaffa26cfb0c247da573fe1a818, thanks @vthib. Removed the line that made YR_RULES opaque from `build.rs` and implemented a `RuleIterator` in the same spirit as the existing Iterators in the code. 

@Hugal31 please let me know whether further changes are required.

Hugal31 commented on src/internals/rules.rs:
```diff
@@ -114,6 +124,63 @@ impl<'a> From<(&'a yara_sys::YR_SCAN_CONTEXT, &'a yara_sys::YR_RULE)> for Rule<'
+        let rule = unsafe { *self.head };
+        let mut result: Option<Self::Item> = None;
+
+        if rule.flags & rule.flags == yara_sys::RULE_FLAGS_NULL {
```
`rule.flags & rules.flags`? Did you meant `(rule.flags &  yara_sys::RULE_FLAGS_NULL) != 0`?

jpohls1 replied:
Yes, good catch, fixed.

Hugal31 commented on yara-sys/src/lib.rs:
```diff
@@ -2,6 +2,8 @@
 #![allow(non_snake_case)]
 #![allow(non_upper_case_globals)]
 
+pub const RULE_FLAGS_NULL: i32 = 0x04;
```
This should be generated by bindgen

Hugal31 commented on src/internals/rules.rs:
```diff
@@ -88,8 +100,8 @@ where
         })
 }
 
-impl<'a> From<(&'a yara_sys::YR_SCAN_CONTEXT, &'a yara_sys::YR_RULE)> for Rule<'a> {
-    fn from((context, rule): (&'a yara_sys::YR_SCAN_CONTEXT, &'a yara_sys::YR_RULE)) -> Self {
+impl<'a> From<&'a yara_sys::YR_RULE> for Rule<'a> {
+    fn from(rule: &'a yara_sys::YR_RULE) -> Self {
```
Can you make this return a `RulesetRule` instead? So the `impl From<...> for Rule` depends on RulesetRule and not the opposite.

jpohls1 replied:
For this, the signature would need to be changed to
```rust
 fn from(rule: &'a mut yara_sys::YR_RULE)
```
because the `inner` pointer of the `RulesetRule` needs to be mutable. Doing that, however, would also require changes in the ` MetadataIterator::from(rule)` and ` TagIterator::from(rule)` since they expect the rule to be immutable.

Moreover, when changing those Iterators too, this leads to a compiler error later:

```
error[E0382]: use of moved value: `rule`
   --> src/internals/rules.rs:112:38
    |
104 |     fn from(rule: &'a mut yara_sys::YR_RULE) -> Self {
    |             ---- move occurs because `rule` has type `&mut YR_RULE`, which does not implement the `Copy` trait
...
111 |         let metadatas = MetadataIterator::from(rule).map(Metadata::from).collect();
    |                                                ---- value moved here
112 |         let tags = TagIterator::from(rule)
    |                                      ^^^^ value used here after move
```

Hugal31 reviewed (changes requested):
You'll have to include the generated binding files (look at the [bindings](https://github.com/Hugal31/yara-rust/actions/runs/7004628232) Github action).

I will probably rename Rule to MatchingRule later (with a deprecation period for Rule).

jpohls1 commented:
@Hugal31 could you please release a new version that includes this change?
